articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md

```diff
@@ -0,0 +1,832 @@
+### Introduction
+
+In recent times, real-time chat applications have grown tremendously. Most companies have adopted them for quick communication. For security purposes, the messages exchanged over the network have to be encrypted. This means, all the messages on transit over the network should be encrypted. 
+
+If a malicious program tries to tap the messages exchanged across the network illegally, the intercepted message would be in encrypted form thus the content of the message will not be compromised.
+
+This article will create a simple chat application where the exchanged messages will have [end-to-end encryption](https://www.preveil.com/blog/end-to-end-encryption/) using the same [secret key](https://www.sciencedirect.com/topics/engineering/secret-key).
+
+### Prerequisites
+
+- Working knowledge of [Node.js](https://nodejs.dev/learn), [Express](https://www.guru99.com/node-js-express.html), [React.js](https://reactjs.org/tutorial/tutorial.html), [aes256 encryption](https://spanning.com/blog/aes-encryption/), and [Socket.io](https://socket.io/docs/v4/index.html).
+
+- A text editor, preferably [VS Code Editor](https://code.visualstudio.com/download).
+
+- A working web browser for testing, preferably [Google Chrome](https://www.google.com/chrome/).
+
+- [Node.js](https://nodejs.org/en/download/) Installed.
+
+### How the App works
+
+We'll create a secret key and store it in the frontend for demonstration purposes. The key is saved in the server's [.ENV](https://create-react-app.dev/docs/adding-custom-environment-variables/) variable where the frontend has been deployed in the production environment.
+
+Whenever a user sends or receives a message, it will be encrypted or decrypted using an [aes256](https://www.npmjs.com/package/aes256) npm package with the same secret key.
+
+### Creating the Backend
+
+For the backend, we'll use Node.js with its framework Express. Socket.io is also be needed to provide real-time, bi-directional communication between the server and the client.
+
+The folder structure for our backend will look as follows:
+
+![backend folder structure](/engineering-education/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/backend-folders.PNG)
+
+### Coding the Backend
+
+**Step 1**
+
+Create a server directory with the name `chatbackend`, browse to the directory, and initialize the server project by running the below commands in the terminal:
+
+```bash
+$ mkdir chatbackend
+$ cd chatbackend
+$ npm init –y
+```
+
+The `package.json` file will be auto-generated, saving the dependencies list that our server will require.
+
+**Step 2**
+
+Next will be to install the dependencies our server will require by running the following commands:
+
+```bash
+$ npm i socket.io express cors colors
+$ npm i -D nodemon
+```
+
+Then, we code the file `dummyuser.js` as below:
```

Comment:
Before you ask the reader to create a file and add the code below, give a brief one-line intro on what the file will contain and will be used for.
Then, you can explain it in detail.